[JianXi] ip

ShiJianXi opened the pull request:
### ChitChatBot

> "Your mind is for having ideas, not holding them." - David Allen

ChitChatBot is a text-based task management tool designed to help you store and organise your thoughts and tasks effortlessly. It is

* **text-based**
* **easy to learn**
* ~~FAST~~ _**SUPER FAST**_ to use

All you need to do is,

1. **download** the .JAR from the [latest release](https://github.com/ShiJianXi/ip/releases/tag/Level-9)
2. open a terminal
3. navigate to the folder (that contains the .JAR file) using `cd` command
4. run the command: 
```bash
java -jar chitchatbot.jar
```

And it is **FREE!** 😉 🆓 

Features:

- [x] Add tasks (**todo, event, deadline**)
- [x] Save tasks locally
- [x] List task
- [x] Managing task (**mark, unmark, delete**)
- [x] Find/Search for a task

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public static void main(String[] args) throws MissingParameterException {
        Path path = Paths.get("data", "chat.txt");
        ChitChatBot chitChatBot = new ChitChatBot(path);
        chitChatBot.run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ShiJianXi committed:
Complete level-0

ShiJianXi committed:
Complete level-1

ShiJianXi committed:
Add line and indentation

ShiJianXi committed:
Implement activity as OOP

ShiJianXi committed:
Complete level-2 and removed unnecessary fields in activity

ShiJianXi committed:
Complete level-3

ShiJianXi committed:
Rename activity to task

ShiJianXi committed:
Add todo

ShiJianXi committed:
Add a printchat method

ShiJianXi committed:
Add deadline.java

ShiJianXi committed:
Fix error in deadline query and add event.java

ShiJianXi committed:
Fix bug in event query

ShiJianXi committed:
Done level-4

ShiJianXi committed:
Complete semi-automatic testing up to level-4

ShiJianXi committed:
Add createtodo in todo.java

ShiJianXi committed:
Add createevent in event.java

ShiJianXi committed:
Add createdeadline in deadline.java

ShiJianXi committed:
Add emptyparametereception for todo queries

ShiJianXi committed:
Fix bug in todo queries

ShiJianXi committed:
Add exception testing for todo into a-textuitest

ShiJianXi committed:
Handle wrong query

ShiJianXi committed:
Update a-textuitest to test for wrong query input

ShiJianXi committed:
Catch indexoutofbound exception for mark queries

ShiJianXi committed:
Update a-textui-test to test for exception handling for mark query

ShiJianXi committed:
Catch numberformationexception for mark queries

ShiJianXi committed:
Add error to the error message for numberformationexception for mark queries

ShiJianXi committed:
Update a-textuitest to test for numberformatexception for mark queries

ShiJianXi committed:
Add exception handling for unmark queries

ShiJianXi committed:
Make exception handling for mark and unmark more specific

ShiJianXi committed:
Update a-textuitest to test for exception handling for unmark

ShiJianXi committed:
Add exception handling for missing parameters for mark and unmark

ShiJianXi committed:
Update a-textuitest to test for missing parameters for mark and unmark

ShiJianXi committed:
Add alreadymarkedexception.java to handle when a task is already mark

ShiJianXi committed:
Update a-textuitest to test for already marked exception

ShiJianXi committed:
Handle exception whereby the users tries to unmark a task that is not yet marked

ShiJianXi committed:
Update a-textuitest to test to handle the already marked exception

ShiJianXi committed:
Rename emptyparameterexception to missingparameterexception

ShiJianXi committed:
Add exception handling for missing parameters for deadline queries

ShiJianXi committed:
Update a-textuitest to test for missing parameters for deadline queries

ShiJianXi committed:
Update a-textuitest

ShiJianXi committed:
Handle the case when there is missing description for deadline query

ShiJianXi committed:
Update a-textuitest to test for missing description for deadline queries

ShiJianXi committed:
Handle missing parameter exception for event queries

ShiJianXi committed:
Update a-textuitest to test for missing parameter exception for event queries

ShiJianXi committed:
Complete level-5 and code cleaned up

ShiJianXi committed:
Add delete method to delete a task

ShiJianXi committed:
Update a-textuitest to test the delete queries

ShiJianXi committed:
Add exception handling for delete queries

ShiJianXi committed:
Handle the exception when user enter too many arguments for delete queries

ShiJianXi committed:
Update a-textuitest to test for delete exceptions

ShiJianXi committed:
Add enum action

ShiJianXi committed:
Complete a-enums and clean up some codes

ShiJianXi committed:
Add data folder to store chat.txt

ShiJianXi committed:
Remove chat.txt

ShiJianXi committed:
Add chat.txt to user if it does not exist

ShiJianXi committed:
Move data folder to root directory

ShiJianXi committed:
Allow data directory to be created as well

ShiJianXi committed:
Modify exception message from deadline to event

ShiJianXi committed:
Modify deadline to event due to previous change

ShiJianXi committed:
Append task to chat.txt for event todo and deadline

ShiJianXi committed:
Allow list query to list from chat.txt

ShiJianXi committed:
Allow task to be tag with a unique id in chat.txt

ShiJianXi committed:
Fix listTask method to list correctly

ShiJianXi committed:
Add method markTest

ShiJianXi committed:
Add unmarkTask method

ShiJianXi committed:
Rename and move markTask and unmarkTask

ShiJianXi committed:
Modify deleteTask in Task.java to delete from chat.txt

ShiJianXi committed:
Complete Level-7

ShiJianXi committed:
Clean up ChitChatBot.java and handle exception for mark within Task.java

ShiJianXi committed:
Handle exception for unmark within Task.java

ShiJianXi committed:
Modify to handle todo query within Todo.java

ShiJianXi committed:
Handle deadline exception within deadline.java

ShiJianXi committed:
Handle event query exception within Event.java

ShiJianXi committed:
Handle delete exception within Task.java

ShiJianXi committed:
Code clean up

ShiJianXi committed:
Merge branch 'branch-Level-7'

ShiJianXi committed:
Change deadline to take in date/time in the date/time format

ShiJianXi committed:
Catch date time exception

ShiJianXi committed:
Update text-ui-test to test for the format for date time

ShiJianXi committed:
Add date time format for Event.java

ShiJianXi committed:
Fix bug in Event.java and Deadline.java

ShiJianXi committed:
Update text-ui-test

ShiJianXi committed:
Merge branch 'branch-Level-8'

ShiJianXi committed:
Add storage class to handle storage related command

ShiJianXi committed:
Add ui class

ShiJianXi committed:
Add parser class to parse user input

ShiJianXi committed:
Complete A-MoreOOP

ShiJianXi committed:
Merge tag 'A-MoreOOP'

ShiJianXi committed:
Complete A-Packages

ShiJianXi committed:
Package into task.deadline

ShiJianXi committed:
Amend package into textuitest

ShiJianXi committed:
Merge branch 'A-Packages'

ShiJianXi committed:
Merge branch 'A-Gradle'

ShiJianXi committed:
Implement DeadlineTest.java to test for methods in Deadline.java

ShiJianXi committed:
Fix bug in text-ui-test

ShiJianXi committed:
Implement TodoTest.java to test for methods in Todo.java

ShiJianXi committed:
Implement EventTest.java to test for method in Event.java

ShiJianXi committed:
Update text-ui-test

ShiJianXi committed:
Add TaskTest and fix bugs in other test

ShiJianXi committed:
Implement delete task after each test

ShiJianXi committed:
Update textuitest

ShiJianXi committed:
Change error to be more specific

ShiJianXi committed:
Merge tag 'A-JUnit'

ShiJianXi committed:
Fix bug in TaskTest.java

ShiJianXi committed:
Implement exit for bye command

ShiJianXi committed:
Adjust seedu.duke to chitchatbot.ChitChatBot

ShiJianXi committed:
Complete A-JavaDoc

ShiJianXi committed:
Do JavaDoc for storage.java

ShiJianXi committed:
Complete A-CodingStandard

ShiJianXi committed:
Implement find command

ShiJianXi committed:
Update text-ui-test to test for Find command

ShiJianXi committed:
Implement command.Find.java

ShiJianXi committed:
Merge branch 'branch-A-JavaDoc'

ShiJianXi committed:
Remove conflict due to merge with Branch-A-CodingStandard

ShiJianXi committed:
Merge branch 'branch-Level-9'

ShiJianXi committed:
Implement FindTest to test for method in Find.java

ShiJianXi committed:
Do JavaDoc for Find.java

ShiJianXi committed:
Adjust JavaDoc to comply to style guide

ShiJianXi committed:
Code clean up

ShiJianXi committed:
Add header comment for class

ShiJianXi committed:
Implement checkstyle using gradle

ShiJianXi committed:
Merge branch 'branch-A-CheckStyle'

ShiJianXi committed:
Implement GUI

ShiJianXi committed:
Code clean up and enhance UI

ShiJianXi committed:
Minor ui fix

ShiJianXi committed:
Enhance GUI

ShiJianXi committed:
Code clean up

ShiJianXi committed:
Implement exit after bye command

ShiJianXi committed:
Update test case

ShiJianXi committed:
Merge branch 'branch-Level-10'

JuhromeAlexander commented on src/main/java/chitchatbot/command/Parser.java:
```diff
@@ -0,0 +1,129 @@
+        try {
+            this.action = Action.valueOf(this.inputArr[0]);
+        } catch (IllegalArgumentException e) {
+            return "OOPS!!! I'm sorry, but I don't know what that means :-(\n"
+                    + "Please use the correct queries:\n"
+                    + "todo <description>\n"
+                    + "deadline <description> /by <Date/Time>\n"
+                    + "event <description> /from <Date&Time> /to <Date&Time>\n"
+                    + "or list to show all the task";
```
I like how you placed the line breaks to improve the readability of the error message.

JuhromeAlexander commented on src/main/java/chitchatbot/task/Todo.java:
```diff
@@ -0,0 +1,71 @@
+
+        storage.appendToFile(newTask.toString());
+
+        return "Got it. I've added this task:\n"
+                + "  " + newTask + "\n"
+                + "Now you have "
+                + Task.getNoOfActivity() + " tasks in the list.";
```
Should this be extracted out? It seems that this return statement is used across the different Tasks

JuhromeAlexander commented on src/main/java/chitchatbot/task/Deadline.java:
```diff
@@ -0,0 +1,126 @@
+
+    protected LocalDate by;
+    protected LocalTime time;
+    private boolean containTime = false;
```
Perhaps the variable name could be 'hasTime' to ensure better readability for the boolean variable

JuhromeAlexander commented on src/main/java/chitchatbot/command/Find.java:
```diff
@@ -0,0 +1,106 @@
+     * @return ArrayList<String> of all the similar task.
+     */
+    public ArrayList<String> findSimilarTask(ArrayList<String> description) {
+        ArrayList<String> originalDescription = new ArrayList<>(List.copyOf(description));
+        ArrayList<String> result = new ArrayList<>();
```
Perhaps the variable names can be changed to its plural forms to better signify the variable is representing a group of objects

JuhromeAlexander reviewed (commented):
LGTM!

tim0tay commented on src/main/java/chitchatbot/Action.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Enum class for the actions that the program accepts
+ */
+public enum Action {
```
Perhaps enums should be capitalised! Eg. BYE, LIST, etc.

tim0tay commented on src/main/java/chitchatbot/command/Find.java:
```diff
@@ -0,0 +1,106 @@
+     * @throws MissingParameterException If the user's input has mising parameters
+     * @see Ui
+     */
+    public String executeFindCommand(String[] inputArr) throws MissingParameterException {
```
Maybe you could call this findArgs or something more informative instead!

tim0tay commented on src/main/java/chitchatbot/command/Find.java:
```diff
@@ -0,0 +1,106 @@
+        }
+
+        StringJoiner result = new StringJoiner("\n");
+        ArrayList<String> lookingFor = new ArrayList<>();
```
It may not be immediately clear what lookingFor means or represents!

tim0tay commented on src/main/java/chitchatbot/storage/Storage.java:
```diff
@@ -0,0 +1,117 @@
+    public String listTask() {
+        try {
+            Scanner scanner = new Scanner(chatFile);
+            StringJoiner toPrint = new StringJoiner("\n");
```
Maybe call this stringToPrint -> should be a noun!

tim0tay reviewed (commented):
Generally the code is pretty clean and variables are well-named! 👍 Good job :>

ShiJianXi committed:
Fix bug with Jar file

Jar file unable to locate Launcher

Add package name in build.gradle for application

User will now be able to launch and run chitchatbot.jar file

ShiJianXi committed:
Implement A-Assertions

Assume that a new file will be
created during initial launch

Implement java assert in the storage constructor

In the event that the file is not created on initial launch
assert fail will be shown

ShiJianXi committed:
Improve code quality for Find.java

Find class: extract methods from
findSimilarTask and executeFindCommand

Method body is too long resulting in bad code quality

Implement SLAP

Implementing SLAP by refactoring out long method body into various methods

ShiJianXi committed:
Improve code quality for Storage.java

Extract method from
initStorage and appendToFile

Apply SLP to improve code quality

ShiJianXi committed:
Improve code quality for deadline.java

ShiJianXi committed:
Improve code quality for Event.java

ShiJianXi committed:
Improve code quality for Task.java

ShiJianXi committed:
Improve code quality in Todo.java

Perform code clean up
in Task.java and Ui.java

ShiJianXi committed:
Merge pull request #1 from ShiJianXi/branch-A-Assertions

Implement A-Assertions

ShiJianXi committed:
Merge branch 'master' into branch-A-CodeQuality

ShiJianXi committed:
Merge pull request #2 from ShiJianXi/branch-A-CodeQuality

Branch a code quality

ShiJianXi committed:
Implement A-CI

ShiJianXi committed:
Merge branch 'branch-A-CI'

ShiJianXi committed:
Add sample data file

ShiJianXi committed:
Fix check style error

ShiJianXi committed:
Merge pull request #3 from ShiJianXi/branch-FixCIBug

Add sample data file

ShiJianXi committed:
Implement Undo command

ShiJianXi committed:
Implement Junit test for Undo.java

ShiJianXi committed:
Fix checkstyle error

ShiJianXi committed:
Remove buggy Junit test

ShiJianXi committed:
Merge pull request from branch-C-Undo

Branch c undo

ShiJianXi committed:
Fix codestyle

ShiJianXi committed:
Improve GUI

ShiJianXi committed:
Merge pull request #6 from ShiJianXi/branch-A-BetterGui

Improve GUI

ShiJianXi committed:
Handle more date time error

ShiJianXi committed:
Run gradle clean

ShiJianXi committed:
Code clean up

ShiJianXi committed:
Fix checkstyle

ShiJianXi committed:
Merge pull request #7 from ShiJianXi/branch-A-MoreErrorHandling

Handle more date time error

ShiJianXi committed:
To test validate gradle wrapper

ShiJianXi committed:
Fix validate gradle wrapper issue

ShiJianXi committed:
Add more test for Event

ShiJianXi committed:
Merge pull request #8 from ShiJianXi/branch-A-MoreTesting

Add more test for Event

ShiJianXi committed:
Add ui.png to docs

ShiJianXi committed:
Implement case insensitive find

ShiJianXi committed:
Finish user guide

ShiJianXi committed:
Fix format issue README.md

ShiJianXi committed:
Update README.md

ShiJianXi committed:
Update README.md to fix format

ShiJianXi committed:
Update README.md

ShiJianXi committed:
Fix format issue in github page

ShiJianXi committed:
Enhance format for README.md

ShiJianXi committed:
Update jar file link in README.md

ShiJianXi committed:
Fix english error in README.md

ShiJianXi committed:
Update repo readme

ShiJianXi committed:
Fix readme format

ShiJianXi committed:
Update readme
